src/fast_forward.rs

```diff
@@ -0,0 +1,505 @@
+    // singleton_top_layer_v1_1.clsp
+    let mod_hash = tree_hash(a, singleton.program);
+    if mod_hash != hex!("7faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9f") {
+        return Err(Error::NotSingletonModHash);
+    }
+
+    // we can only fast-forward if the coin amount stay the same
+    // this is to minimize the risk of producing an invalid spend, after
+    // fast-forward. e.g. we might end up attempting to spend more that the
+    // amount of the coin
+    if coin.amount != new_solution.lineage_proof.parent_amount || coin.amount != new_parent.amount {
+        return Err(Error::CoinAmountMismatch);
+    }
+
+    // given the parent's parent, the parent's inner puzzle and parent's amount,
+    // we can compute the hash of the curried inner puzzle for our parent coin
+    let parent_puzzle_hash = curry_and_treehash(
+        &new_solution.lineage_proof.parent_inner_puzzle_hash,
+        &singleton.args.singleton_struct,
+    );
+
+    // now that we know the parent coin's puzzle hash, we have all the pieces to
+    // compute the coin being spent (before the fast-forward).
+    let parent_coin = Coin {
+        parent_coin_info: new_solution.lineage_proof.parent_parent_coin_id,
+        puzzle_hash: parent_puzzle_hash,
+        amount: new_solution.lineage_proof.parent_amount,
+    };
+
+    if parent_coin.coin_id() != coin.parent_coin_info {
+        return Err(Error::ParentCoinMismatch);
+    }
+
+    let inner_puzzle_hash = tree_hash(a, singleton.args.inner_puzzle);
+    if inner_puzzle_hash != new_solution.lineage_proof.parent_inner_puzzle_hash {
+        return Err(Error::InnerPuzzleHashMismatch);
+    }
+
+    let puzzle_hash = tree_hash(a, puzzle);
+
+    if puzzle_hash != new_parent.puzzle_hash || puzzle_hash != coin.puzzle_hash {
+        // we can only fast-forward if the puzzle hash match the new coin
+        // the spend is assumed to be valied already, so we don't check it
+        // against the original coin being spent
+        return Err(Error::PuzzleHashMismatch);
+    }
+
+    // update the solution to use the new parent coin's information
+    new_solution.lineage_proof.parent_parent_coin_id = new_parent.parent_coin_info;
+
+    let expected_new_parent = new_parent.coin_id();
+
+    if new_coin.parent_coin_info != expected_new_parent {
+        return Err(Error::CoinMismatch);
+    }
+
+    Ok(new_solution.to_clvm(a)?)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::gen::run_puzzle::run_puzzle;
+    use chia_protocol::CoinSpend;
+    use chia_traits::streamable::Streamable;
+    use clvmr::serde::{node_from_bytes, node_to_bytes};
+    use rstest::rstest;
+    use std::fs;
+    use std::io::Cursor;
+
+    // this test loads CoinSpends from file (Coin, puzzle, solution)-triples
+    // and "fast-forwards" the spend onto a few different parent-parent coins
+    // and ensures the spends are still valid
+    #[rstest]
+    #[case("e3c0")]
+    #[case("bb13")]
```

Comment:
these are filenames of actual singleton spends I pulled out of the mainnet chain